pkg/operation/botanist/waiter.go

```diff
@@ -325,6 +326,67 @@ func (b *Botanist) WaitUntilNodesDeleted(ctx context.Context) error {
 	})
 }
 
+// WaitUntilNoPodRunning waits until there is no running Pod in the shoot cluster.
+func (b *Botanist) WaitUntilNoPodRunning(ctx context.Context) error {
+	b.Logger.Info("waiting until there are no running Pods in the shoot cluster...")
+
+	return retry.Until(ctx, 5*time.Second, func(ctx context.Context) (done bool, err error) {
+		podList := &corev1.PodList{}
+		if err := b.K8sShootClient.Client().List(ctx, podList); err != nil {
+			return retry.SevereError(err)
+		}
+
+		for _, pod := range podList.Items {
+			if pod.Status.Phase == corev1.PodRunning {
+				msg := fmt.Sprintf("waiting until there are no running Pods in the shoot cluster... "+
+					"there is still at least one running Pod in the shoot cluster: %s/%s", pod.Namespace, pod.Name)
+				b.Logger.Info(msg)
+				return retry.MinorError(fmt.Errorf(msg))
+			}
+		}
+
+		return retry.Ok()
+	})
+}
+
+// WaitUntilEndpointsDoNotContainPodIPs waits until all endpoints in the shoot cluster to not contain any IPs from the Shoot's PodCIDR.
+func (b *Botanist) WaitUntilEndpointsDoNotContainPodIPs(ctx context.Context) error {
+	b.Logger.Info("waiting until there are no Endpoints containing Pod IPs in the shoot cluster...")
+
+	var podsNetwork *net.IPNet
+	if val := b.Shoot.Info.Spec.Networking.Pods; val != nil {
+		var err error
+		_, podsNetwork, err = net.ParseCIDR(*val)
+		if err != nil {
+			return fmt.Errorf("unable to check if there are still Endpoints containing Pod IPs in the shoot cluster. Shoots's Pods network could not be parsed: %+v", err)
+		}
+	} else {
+		return fmt.Errorf("unable to check if there are still Endpoints containing Pod IPs in the shoot cluster. Shoot's Pods network is empty")
+	}
+
+	return retry.Until(ctx, 5*time.Second, func(ctx context.Context) (done bool, err error) {
+		endpointsList := &corev1.EndpointsList{}
+		if err := b.K8sShootClient.Client().List(ctx, endpointsList); err != nil {
+			return retry.SevereError(err)
+		}
+
+		for _, endpoints := range endpointsList.Items {
```

Comment:
It probably makes sense to check the endpoints in parallel, i.e. one go routine per endpoint. WDYT?